From 1765472b13c7f1e81892cad2c09afb1876072e17 Mon Sep 17 00:00:00 2001
From: ran shalit <rshalit@habana.ai>
Date: Sun, 18 Apr 2021 22:05:51 +0300
Subject: [PATCH] [SW-39246] spi: add erase capability

Add erase capability in spi flash, as a preparation
for a later patch which handles request command (fit)
for flash erase. Also required increase of size as the
spi code is common from ppboot as well.
Erase timeout was measured (about ~500msec), and
requires increase of wait timeout for commands.

Change-Id: I90839a1069cbd81dc4a1daa1f8e038e528b45bc8
Signed-off-by: ran shalit <rshalit@habana.ai>
---

diff --git a/common/spi.c b/common/spi.c
index 7f0ccf1..e18ecb1 100644
--- a/common/spi.c
+++ b/common/spi.c
@@ -34,11 +34,16 @@
 #define CMD_WRITE_DISABLE		0x04
 #define CMD_READ_STATUS			0x05
 #define CMD_WRITE_ENABLE		0x06
+#define CMD_FLAG_STATUS			0x70
+#define CMD_ERASE			0xD8
+#define READ_STATUS_PEC			(1 << 7)
 #define frame_size_bits(b)		((b) << 16)
 
 /* Used for Macronix flashes only. */
 #define SPINOR_OP_ENSO			0xB1
 #define SPINOR_OP_EXSO			0xC1
+
+#define PAGE_SIZE			0x10000
 
 static unsigned int spi_regs;
 /*
@@ -139,8 +144,20 @@
 	spi_release_bus();
 }
 
+static void spi_cmd_write(unsigned int cmd, int cmd_size)
+{
+	int i, bit_shift = 0;
+
+	for (i = 0; i < cmd_size; i++) {
+		while (!spi_fifo_is_empty(spi_regs + TXFLR))
+			;
+		regmap_writeb((cmd >> bit_shift) & 0xff, spi_regs + DR0);
+		bit_shift += 8;
+	}
+}
+
 static void spi_special_cmd_csel(unsigned int cmd, unsigned char *val,
-								size_t size)
+				 size_t size, int cmd_size)
 {
 	int i;
 
@@ -148,35 +165,36 @@
 
 	spi_dev_update_frame_size(sizeof(*val));
 
-	while (!spi_fifo_is_empty(spi_regs + TXFLR))
-		;
-	regmap_writeb(cmd, spi_regs + DR0);
-	for (i = 0; i < size; i++) {
-		while (!spi_fifo_is_empty(spi_regs + TXFLR))
-			;
-		regmap_writeb(val[i], spi_regs + DR0);
-	}
+	spi_cmd_write(cmd, cmd_size);
 
-	while (spi_fifo_is_empty(spi_regs + RXFLR))
-		;
-	regmap_readb(spi_regs + DR0);
-	for (i = 0; i < size; i++) {
+	if (val) {
+		for (i = 0; i < size; i++) {
+			while (!spi_fifo_is_empty(spi_regs + TXFLR))
+				;
+			regmap_writeb(val[i], spi_regs + DR0);
+		}
+
 		while (spi_fifo_is_empty(spi_regs + RXFLR))
 			;
 		regmap_readb(spi_regs + DR0);
+		for (i = 0; i < size; i++) {
+			while (spi_fifo_is_empty(spi_regs + RXFLR))
+				;
+			regmap_readb(spi_regs + DR0);
+		}
 	}
-
 	spi_release_bus();
 }
 
 
 /*
  * 40ms as worst case WRSR cycle time given in mx25u25635f datasheet.
- * Kept some extra(10ms) as timeout (total 50ms).
+ * 2sec is the worse erase time according to DS.
  */
-#define SPI_WRITE_WAIT_TIMEOUT_US	50000
-static void spi_common_write(unsigned int cmd, unsigned char *buf, size_t size)
+#define SPI_WRITE_WAIT_TIMEOUT_US	2000000
+static int spi_common_write(unsigned int cmd, unsigned char *buf, size_t size, int cmd_size)
 {
+	int ret = 0;
 	unsigned char sr;
 	u64 start = timestamp_get_us();
 
@@ -184,18 +202,22 @@
 	do {
 		spi_char_cmd_csel(CMD_READ_STATUS, &sr);
 		if (timeout_expired(start, SPI_WRITE_WAIT_TIMEOUT_US))
-			return;
+			return 1;
 	} while (!(sr & READ_STATUS_WEL));
 
 	start = timestamp_get_us();
-	spi_special_cmd_csel(cmd, buf, size);
+	spi_special_cmd_csel(cmd, buf, size, cmd_size);
 	do {
 		spi_char_cmd_csel(CMD_READ_STATUS, &sr);
-		if (timeout_expired(start, SPI_WRITE_WAIT_TIMEOUT_US))
+		if (timeout_expired(start, SPI_WRITE_WAIT_TIMEOUT_US)) {
+			ret = 1;
 			break;
+		}
 	} while (sr & READ_STATUS_WIP);
 
 	spi_char_cmd_csel(CMD_WRITE_DISABLE, NULL);
+
+	return ret;
 }
 
 static void spi_read(void *dst, unsigned int offset, unsigned int size)
@@ -210,6 +232,21 @@
 		spi_dev_dword_cmd(0, dst + i);
 	spi_dev_dword_cmd(0, NULL);
 	spi_release_bus();
+}
+
+int spi_erase(unsigned int offset, unsigned int size)
+{
+	unsigned int i;
+	unsigned int cmd = CMD_ERASE;
+	int ret;
+
+	for (i = 0; i < size; i += PAGE_SIZE) {
+		ret = spi_common_write(cmd | __builtin_bswap32(offset + i), NULL, 0, sizeof(cmd));
+		if (ret)
+			return ret;
+	}
+
+	return 0;
 }
 
 __weak void spi_base_get(unsigned int *regs)
@@ -243,7 +280,7 @@
 
 	spi_char_cmd_csel(CMD_READ_STATUS, &sr);
 	sr |= PROT_MASK;
-	spi_common_write(CMD_WRITE_STATUS, &sr, sizeof(sr));
+	spi_common_write(CMD_WRITE_STATUS, &sr, sizeof(sr), 1);
 	sr = 0; /* reset old value before read */
 	spi_char_cmd_csel(CMD_READ_STATUS, &sr);
 
diff --git a/include/spi.h b/include/spi.h
index 55a31bc..9d9564b 100644
--- a/include/spi.h
+++ b/include/spi.h
@@ -31,5 +31,6 @@
 void spi_copy_otp(void *dst, unsigned int offset, unsigned int size);
 unsigned int get_cs_pad_sel(void);
 void spi_prot(void);
+int spi_erase(unsigned int offset, unsigned int size);
 
 #endif /* _HL_SPI_H_ */
